fuzzers/main.py

```diff
@@ -0,0 +1,642 @@
+        threading.Thread.__init__(self, daemon=True)
+        self.crate = crate
+        self.runner = runner
+        self.dir = dir
+        self.bucket = bucket
+        self.version = version
+        self.path = path
+        self.log_file = log_file
+        self.is_artifacts = is_artifacts
+
+        self.exit_event = threading.Event()
+        self.corpus_uploaded_metric = FUZZ_CORPUS_UPLOADED.labels(crate, runner)
+        self.corpus_deleted_metric = FUZZ_CORPUS_DELETED.labels(crate, runner)
+        self.artifacts_found_metric = FUZZ_ARTIFACTS_FOUND.labels(crate, runner)
+
+    def run(self):
+        utils.mkdirs(self.dir / self.path)
+        i = inotify.adapters.Inotify()
+        i.add_watch((self.dir / self.path).as_posix())
+        while not self.exit_event.is_set():
+            # check for exit_event at most every second
+            for event in i.event_gen(yield_nones = False, timeout_s = 1):
+                if self.exit_event.is_set():
+                    break
+                (_, event_types, _, filename) = event
+                local_filename = self.dir / self.path / filename
+                remote_filename = f'{self.version}/{self.path}/{filename}'
+
+                if 'IN_CLOSE_WRITE' in event_types:
+                    self.log_file.write(
+                        f"Uploading new corpus item {local_filename} to GCS {remote_filename}\n"
+                    )
+                    self.log_file.flush()
+                    try:
+                        # TODO: batch uploads
+                        self.bucket.blob(remote_filename).upload_from_filename(local_filename)
+                        if self.is_artifacts:
+                            self._report_artifact(remote_filename, self.crate, self.runner)
+                        else:
+                            self.corpus_uploaded_metric.inc()
+                    except FileNotFoundError:
+                        pass # Ignore, as it'd mean the file has been deleted already
+
+                if 'IN_DELETE' in event_types and not self.is_artifacts:
+                    self.log_file.write(
+                        f"Removing now-removed corpus item {local_filename} as GCS {remote_filename}\n"
+                    )
+                    self.log_file.flush()
+                    try:
+                        # TODO: batch
+                        self.bucket.blob(remote_filename).delete()
+                        self.corpus_deleted_metric.inc()
+                    except google.api_core.exceptions.NotFound:
+                        pass # Ignore, as it'd mean the file isn't there already
+
+    def _report_artifact(self, gcs_path: str, crate: str, runner: str) -> None:
+        self.artifacts_found_metric.inc()
+
+class FuzzProcess:
+    def __init__(
+        self,
+        corpus_vers: str,
+        branch: BranchType,
+        target: TargetType,
+        repo_dir: pathlib.Path,
+        log_path: pathlib.Path,
+        log_filepath: pathlib.Path,
```

Comment:
`log_path` isn’t really needed here though, is it? It’s always `log_filepath.parent`, no? Also, wouldn’t we want to print full `log_filepath` in log messages rather than just the directory path such that we could get rid of `self.log_path`?